storage_proxy: handler responses, use pointers to default constructed values

gusev-p opened the pull request:
The Seastar RPC infrastructure does not support null values in tuples in handler responses. In this commit we add a function `make_default_rpc_tuple` which replaces them with pointers to default constructed values.

The problem was introduced in [this commit](https://github.com/scylladb/scylladb/commit/2d791a5ed4de5945fab80577266dbbbdef113c77). The function `encode_replica_exception_for_rpc` used `default_tuple_maker` callback to create tuples containing exceptions. Callers returned pointers to default initialised classes in this callback, e.g. `foreign_ptr(make_lw_shared<reconcilable_result>())`. The commit changed this to just `SourceTuple{}`, which means nullptr for pointer types.

Fixes: #14282

avikivity commented:
Can you give a higher level explanation? Maybe the solution should be to support null values? It's hard to say anything concrete because the description is so low-level.

gleb-cloudius commented:
As far as I understand this fixes regression introduced by https://github.com/scylladb/scylladb/commit/2d791a5ed4de5945fab80577266dbbbdef113c77 Can you please explain what exactly the regression was and how this fixes it.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1999/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2003/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2018/

bhalevy commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2018/

https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2018/consoleText
```
[2023-06-23T10:19:49.848Z] FAILED: build/dev/utils/tuple_utils.hh.o 
[2023-06-23T10:19:49.848Z] clang++ -MD -MT build/dev/utils/tuple_utils.hh.o -MF build/dev/utils/tuple_utils.hh.o.d -I/jenkins/workspace/scylla-master/scylla-ci/scylla/seastar/include -I/jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/gen/include -U_FORTIFY_SOURCE -DSEASTAR_SSTRING -Werror=unused-result -fstack-clash-protection -DSEASTAR_API_LEVEL=7 -DSEASTAR_BUILD_SHARED_LIBS -DSEASTAR_ENABLE_ALLOC_FAILURE_INJECTION -DSEASTAR_SCHEDULING_GROUPS_COUNT=16 -DSEASTAR_TYPE_ERASE_MORE -DFMT_SHARED -DSEASTAR_BROKEN_SOURCE_LOCATION -I/usr/include/p11-kit-1 --std=gnu++20  -ffile-prefix-map=/jenkins/workspace/scylla-master/scylla-ci/scylla=. -march=westmere  -DBOOST_TEST_DYN_LINK   -DNOMINMAX -DNOMINMAX -fvisibility=hidden  -Wall -Werror -Wno-mismatched-tags -Wno-tautological-compare -Wno-c++11-narrowing -Wno-ignored-attributes -Wno-overloaded-virtual -Wno-unused-command-line-argument -Wno-unsupported-friend -Wno-delete-non-abstract-non-virtual-dtor -Wno-braced-scalar-init -Wno-implicit-int-float-conversion -Wno-delete-abstract-non-virtual-dtor -Wno-psabi -Wno-narrowing -Wno-nonnull -Wno-uninitialized -Wno-error=deprecated-declarations -DXXH_PRIVATE_API -DSEASTAR_TESTING_MAIN -DFMT_DEPRECATED_OSTREAM   -ffile-prefix-map=/jenkins/workspace/scylla-master/scylla-ci/scylla=. -march=westmere -DDEVEL -DSEASTAR_ENABLE_ALLOC_FAILURE_INJECTION -DSCYLLA_ENABLE_ERROR_INJECTION -O2 -DSCYLLA_BUILD_MODE=dev -iquote. -iquote build/dev/gen  --include utils/tuple_utils.hh -c -o build/dev/utils/tuple_utils.hh.o build/dev/gen/empty.cc
[2023-06-23T10:19:49.848Z] In file included from <built-in>:1:
[2023-06-23T10:19:49.848Z] ./utils/tuple_utils.hh:62:39: error: use of undeclared identifier 'lw_shared_ptr'
[2023-06-23T10:19:49.848Z]             static auto create(marker<lw_shared_ptr<T>>) {
[2023-06-23T10:19:49.848Z]                                       ^
[2023-06-23T10:19:49.848Z] ./utils/tuple_utils.hh:66:39: error: use of undeclared identifier 'shared_ptr'
[2023-06-23T10:19:49.848Z]             static auto create(marker<shared_ptr<T>>) {
[2023-06-23T10:19:49.848Z]                                       ^
[2023-06-23T10:19:49.848Z] ./utils/tuple_utils.hh:70:39: error: use of undeclared identifier 'foreign_ptr'
[2023-06-23T10:19:49.848Z]             static auto create(marker<foreign_ptr<T>>) {
[2023-06-23T10:19:49.848Z]                                       ^
```

bhalevy reviewed (approved):
modulo the check header failure this change lgtm.
It was better if we could save the construction of the default tuple and just reply with the exception, but it might be an overkill in this case.

avikivity commented:
> > Can you give a higher level explanation?
> 
> I elaborated in the PR description.

Thanks, now I understand.

> 
> > Maybe the solution should be to support null values?
> 
> That would be great, but as [discussed here](https://github.com/scylladb/scylladb/issues/14282#issuecomment-1602617333) pointers are really serialized as values, so the server verb signature with pointers corresponds to the client one without pointers.

Given that creating default constructed values is reasonable. I'd still like to avoid it if the common case is that they're not needed, as this is a hot path, but let's fix the regression first.

avikivity commented:
But, please update the patch changelog. Cover letters for single patches don't make it into git log.

avikivity commented on utils/tuple_utils.hh:
```diff
@@ -11,6 +11,8 @@
 #include <tuple>
+#include <seastar/core/shared_ptr.hh>
+#include <seastar/core/sharded.hh>
 
```
Please move the new code to a new file rpc-utils.hh. The new code isn't about tuples, it's about rpc, and random code that wants to manipulate tuples shouldn't learn about shared_ptr etc.

gusev-p replied:
done

avikivity commented on service/storage_proxy.cc:
```diff
@@ -134,7 +134,7 @@ static future<ResultTuple> encode_replica_exception_for_rpc(gms::feature_service
     if (features.typed_errors_in_read_rpc) {
         if (auto ex = replica::try_encode_replica_exception(eptr); ex) {
-            return make_ready_future<ResultTuple>(utils::tuple_insert<ResultTuple>(SourceTuple{}, std::move(ex)));
+            return make_ready_future<ResultTuple>(utils::tuple_insert<ResultTuple>(utils::make_default_rpc_tuple<SourceTuple>(), std::move(ex)));
```
Given this is an exceptional path, my comment about optimizing it isn't relevant.

gusev-p commented:
> > CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2018/
> 
> https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2018/consoleText
> 
> ```
> [2023-06-23T10:19:49.848Z] [903/905] CHECKHH utils/tuple_utils.hh
> [2023-06-23T10:19:49.848Z] FAILED: build/dev/utils/tuple_utils.hh.o 
> [2023-06-23T10:19:49.848Z] clang++ -MD -MT build/dev/utils/tuple_utils.hh.o -MF build/dev/utils/tuple_utils.hh.o.d -I/jenkins/workspace/scylla-master/scylla-ci/scylla/seastar/include -I/jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/gen/include -U_FORTIFY_SOURCE -DSEASTAR_SSTRING -Werror=unused-result -fstack-clash-protection -DSEASTAR_API_LEVEL=7 -DSEASTAR_BUILD_SHARED_LIBS -DSEASTAR_ENABLE_ALLOC_FAILURE_INJECTION -DSEASTAR_SCHEDULING_GROUPS_COUNT=16 -DSEASTAR_TYPE_ERASE_MORE -DFMT_SHARED -DSEASTAR_BROKEN_SOURCE_LOCATION -I/usr/include/p11-kit-1 --std=gnu++20  -ffile-prefix-map=/jenkins/workspace/scylla-master/scylla-ci/scylla=. -march=westmere  -DBOOST_TEST_DYN_LINK   -DNOMINMAX -DNOMINMAX -fvisibility=hidden  -Wall -Werror -Wno-mismatched-tags -Wno-tautological-compare -Wno-c++11-narrowing -Wno-ignored-attributes -Wno-overloaded-virtual -Wno-unused-command-line-argument -Wno-unsupported-friend -Wno-delete-non-abstract-non-virtual-dtor -Wno-braced-scalar-init -Wno-implicit-int-float-conversion -Wno-delete-abstract-non-virtual-dtor -Wno-psabi -Wno-narrowing -Wno-nonnull -Wno-uninitialized -Wno-error=deprecated-declarations -DXXH_PRIVATE_API -DSEASTAR_TESTING_MAIN -DFMT_DEPRECATED_OSTREAM   -ffile-prefix-map=/jenkins/workspace/scylla-master/scylla-ci/scylla=. -march=westmere -DDEVEL -DSEASTAR_ENABLE_ALLOC_FAILURE_INJECTION -DSCYLLA_ENABLE_ERROR_INJECTION -O2 -DSCYLLA_BUILD_MODE=dev -iquote. -iquote build/dev/gen  --include utils/tuple_utils.hh -c -o build/dev/utils/tuple_utils.hh.o build/dev/gen/empty.cc
> [2023-06-23T10:19:49.848Z] In file included from <built-in>:1:
> [2023-06-23T10:19:49.848Z] ./utils/tuple_utils.hh:62:39: error: use of undeclared identifier 'lw_shared_ptr'
> [2023-06-23T10:19:49.848Z]             static auto create(marker<lw_shared_ptr<T>>) {
> [2023-06-23T10:19:49.848Z]                                       ^
> [2023-06-23T10:19:49.848Z] ./utils/tuple_utils.hh:66:39: error: use of undeclared identifier 'shared_ptr'
> [2023-06-23T10:19:49.848Z]             static auto create(marker<shared_ptr<T>>) {
> [2023-06-23T10:19:49.848Z]                                       ^
> [2023-06-23T10:19:49.848Z] ./utils/tuple_utils.hh:70:39: error: use of undeclared identifier 'foreign_ptr'
> [2023-06-23T10:19:49.848Z]             static auto create(marker<foreign_ptr<T>>) {
> [2023-06-23T10:19:49.848Z]                                       ^
> ```

TBH, I don't understand the failure. It happens only in dev mode and doesn't reproduce locally. All the needed `@include`s are in place.

gleb-cloudius commented:
> 
> TBH, I don't understand the failure. It happens only in dev mode and doesn't reproduce locally. All the needed `@include`s are in place.

It fails while running `ninja checkheaders` which usually nobody runs locally.

gusev-p commented:
> > > CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2018/
> > 
> > 
> > https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2018/consoleText
> > ```
> > [2023-06-23T10:19:49.848Z] [903/905] CHECKHH utils/tuple_utils.hh
> > [2023-06-23T10:19:49.848Z] FAILED: build/dev/utils/tuple_utils.hh.o 
> > [2023-06-23T10:19:49.848Z] clang++ -MD -MT build/dev/utils/tuple_utils.hh.o -MF build/dev/utils/tuple_utils.hh.o.d -I/jenkins/workspace/scylla-master/scylla-ci/scylla/seastar/include -I/jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/gen/include -U_FORTIFY_SOURCE -DSEASTAR_SSTRING -Werror=unused-result -fstack-clash-protection -DSEASTAR_API_LEVEL=7 -DSEASTAR_BUILD_SHARED_LIBS -DSEASTAR_ENABLE_ALLOC_FAILURE_INJECTION -DSEASTAR_SCHEDULING_GROUPS_COUNT=16 -DSEASTAR_TYPE_ERASE_MORE -DFMT_SHARED -DSEASTAR_BROKEN_SOURCE_LOCATION -I/usr/include/p11-kit-1 --std=gnu++20  -ffile-prefix-map=/jenkins/workspace/scylla-master/scylla-ci/scylla=. -march=westmere  -DBOOST_TEST_DYN_LINK   -DNOMINMAX -DNOMINMAX -fvisibility=hidden  -Wall -Werror -Wno-mismatched-tags -Wno-tautological-compare -Wno-c++11-narrowing -Wno-ignored-attributes -Wno-overloaded-virtual -Wno-unused-command-line-argument -Wno-unsupported-friend -Wno-delete-non-abstract-non-virtual-dtor -Wno-braced-scalar-init -Wno-implicit-int-float-conversion -Wno-delete-abstract-non-virtual-dtor -Wno-psabi -Wno-narrowing -Wno-nonnull -Wno-uninitialized -Wno-error=deprecated-declarations -DXXH_PRIVATE_API -DSEASTAR_TESTING_MAIN -DFMT_DEPRECATED_OSTREAM   -ffile-prefix-map=/jenkins/workspace/scylla-master/scylla-ci/scylla=. -march=westmere -DDEVEL -DSEASTAR_ENABLE_ALLOC_FAILURE_INJECTION -DSCYLLA_ENABLE_ERROR_INJECTION -O2 -DSCYLLA_BUILD_MODE=dev -iquote. -iquote build/dev/gen  --include utils/tuple_utils.hh -c -o build/dev/utils/tuple_utils.hh.o build/dev/gen/empty.cc
> > [2023-06-23T10:19:49.848Z] In file included from <built-in>:1:
> > [2023-06-23T10:19:49.848Z] ./utils/tuple_utils.hh:62:39: error: use of undeclared identifier 'lw_shared_ptr'
> > [2023-06-23T10:19:49.848Z]             static auto create(marker<lw_shared_ptr<T>>) {
> > [2023-06-23T10:19:49.848Z]                                       ^
> > [2023-06-23T10:19:49.848Z] ./utils/tuple_utils.hh:66:39: error: use of undeclared identifier 'shared_ptr'
> > [2023-06-23T10:19:49.848Z]             static auto create(marker<shared_ptr<T>>) {
> > [2023-06-23T10:19:49.848Z]                                       ^
> > [2023-06-23T10:19:49.848Z] ./utils/tuple_utils.hh:70:39: error: use of undeclared identifier 'foreign_ptr'
> > [2023-06-23T10:19:49.848Z]             static auto create(marker<foreign_ptr<T>>) {
> > [2023-06-23T10:19:49.848Z]                                       ^
> > ```
> 
> TBH, I don't understand the failure. It happens only in dev mode and doesn't reproduce locally. All the needed `@include`s are in place.

seems that `seastar::lw_shared_ptr` should solve the problem

gusev-p commented:
new version:
  * extracted `rpc_utils.hh`;
  * added `Tuple` concept to the `make_default_rpc_tuple` template parameter;
  * improved the commit message;
  * fixed CI issue (added `seastar::` prefix to smart pointer `utils:: internal:: rpc_default_value::create` overloads).

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2055/

avikivity commented:
> > > CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2018/
> > 
> > 
> > https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2018/consoleText
> > ```
> > [2023-06-23T10:19:49.848Z] [903/905] CHECKHH utils/tuple_utils.hh
> > [2023-06-23T10:19:49.848Z] FAILED: build/dev/utils/tuple_utils.hh.o 
> > [2023-06-23T10:19:49.848Z] clang++ -MD -MT build/dev/utils/tuple_utils.hh.o -MF build/dev/utils/tuple_utils.hh.o.d -I/jenkins/workspace/scylla-master/scylla-ci/scylla/seastar/include -I/jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/gen/include -U_FORTIFY_SOURCE -DSEASTAR_SSTRING -Werror=unused-result -fstack-clash-protection -DSEASTAR_API_LEVEL=7 -DSEASTAR_BUILD_SHARED_LIBS -DSEASTAR_ENABLE_ALLOC_FAILURE_INJECTION -DSEASTAR_SCHEDULING_GROUPS_COUNT=16 -DSEASTAR_TYPE_ERASE_MORE -DFMT_SHARED -DSEASTAR_BROKEN_SOURCE_LOCATION -I/usr/include/p11-kit-1 --std=gnu++20  -ffile-prefix-map=/jenkins/workspace/scylla-master/scylla-ci/scylla=. -march=westmere  -DBOOST_TEST_DYN_LINK   -DNOMINMAX -DNOMINMAX -fvisibility=hidden  -Wall -Werror -Wno-mismatched-tags -Wno-tautological-compare -Wno-c++11-narrowing -Wno-ignored-attributes -Wno-overloaded-virtual -Wno-unused-command-line-argument -Wno-unsupported-friend -Wno-delete-non-abstract-non-virtual-dtor -Wno-braced-scalar-init -Wno-implicit-int-float-conversion -Wno-delete-abstract-non-virtual-dtor -Wno-psabi -Wno-narrowing -Wno-nonnull -Wno-uninitialized -Wno-error=deprecated-declarations -DXXH_PRIVATE_API -DSEASTAR_TESTING_MAIN -DFMT_DEPRECATED_OSTREAM   -ffile-prefix-map=/jenkins/workspace/scylla-master/scylla-ci/scylla=. -march=westmere -DDEVEL -DSEASTAR_ENABLE_ALLOC_FAILURE_INJECTION -DSCYLLA_ENABLE_ERROR_INJECTION -O2 -DSCYLLA_BUILD_MODE=dev -iquote. -iquote build/dev/gen  --include utils/tuple_utils.hh -c -o build/dev/utils/tuple_utils.hh.o build/dev/gen/empty.cc
> > [2023-06-23T10:19:49.848Z] In file included from <built-in>:1:
> > [2023-06-23T10:19:49.848Z] ./utils/tuple_utils.hh:62:39: error: use of undeclared identifier 'lw_shared_ptr'
> > [2023-06-23T10:19:49.848Z]             static auto create(marker<lw_shared_ptr<T>>) {
> > [2023-06-23T10:19:49.848Z]                                       ^
> > [2023-06-23T10:19:49.848Z] ./utils/tuple_utils.hh:66:39: error: use of undeclared identifier 'shared_ptr'
> > [2023-06-23T10:19:49.848Z]             static auto create(marker<shared_ptr<T>>) {
> > [2023-06-23T10:19:49.848Z]                                       ^
> > [2023-06-23T10:19:49.848Z] ./utils/tuple_utils.hh:70:39: error: use of undeclared identifier 'foreign_ptr'
> > [2023-06-23T10:19:49.848Z]             static auto create(marker<foreign_ptr<T>>) {
> > [2023-06-23T10:19:49.848Z]                                       ^
> > ```
> 
> TBH, I don't understand the failure. It happens only in dev mode and doesn't reproduce locally. All the needed `@include`s are in place.

Missing #include "seastarx.hh"

avikivity commented:
Qualifying with seastar:: is fine too.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2056/
